src/psyclone/tests/psyir/nodes/reference_test.py

```diff
@@ -214,3 +214,221 @@ def test_reference_can_be_copied():
+                     "reference_accesses")
+
+
+def test_reference_previous_access(fortran_reader):
+    '''Test the previous_access function for a Reference'''
+    code = '''subroutine my_sub()
+    integer :: a
+    integer :: b
+    a = 1
+    b = 1
+    a = b
+    end subroutine my_sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    b = routine.children[1].lhs
+    a_2 = routine.children[2].lhs
+    b_2 = routine.children[2].rhs
+    assert a.previous_access() is None
+    assert b.previous_access() is None
+    assert a_2.previous_access() is a
+    assert b_2.previous_access() is b
+
+    code = '''subroutine my_sub()
+    integer :: a
+    integer :: b
+    a = 1
+    do a = 0, 10
+        b = a
+    end do
+    end subroutine'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    loop = routine.children[1]
+    b_a = loop.loop_body.children[0].lhs
+    a_2 = loop.loop_body.children[0].rhs
+    assert a.previous_access() is None
+    assert b_a.previous_access() is None
+    assert a_2.previous_access() is loop
+
+    # Check the function for basic structures
+    code = '''subroutine my_sub()
+    type :: x
+       integer :: a
+       real :: b
+    end type
+    type(x) :: i
+    i%a = 2
+    i%b = 0.1
+    i%a = 3
+    i%b = 0.2
+    end subroutine'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    b = routine.children[1].lhs
+    a_2 = routine.children[2].lhs
+    b_2 = routine.children[3].lhs
+    assert a.previous_access() is None
+    assert b.previous_access() is None
+    assert a_2.previous_access() is a
+    assert b_2.previous_access() is b
+
+    # Check the function for array access
+    code = '''subroutine my_sub()
+    integer, dimension(100) :: a
+    a(0) = 2
+    a(1) = 2
+    end subroutine my_sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    a_2 = routine.children[1].lhs
+    assert a.previous_access() is None
+    assert a_2.previous_access() is a
+
+    # Check if statements don't affect
+    code = '''subroutine my_sub()
+   integer :: a
+   logical :: b = .false.
+   a = 1
+   if(b) then
+     a = 0
+   end if
+   a = 2
+   end subroutine'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    a_2 = routine.children[1].if_body.children[0].lhs
+    a_3 = routine.children[2].lhs
+    assert a.previous_access() is None
+    assert a_2.previous_access() is a
+    assert a_3.previous_access() is a_2
+
+
+def test_reference_previous_access_with_codeblock(fortran_reader):
+    code = '''subroutine my_sub()
+    character, dimension(100) :: a
+    write(a, "A") "mytest"
+    a = "test"
+    end subroutine
+    '''
```

Comment:
Could you also test what happens with initialisation expressions, e.g.:
```
module my_mod
   use external_mod, only: a
   contains
   subroutine my_sub()
      integer :: b = a + a +1
      b = a + 3
   end subroutine my_sub
end module my_mod
```
I expect that the 'a' in the statement won't have a previous_access to the initialisation expression (because that's not part of the tree, which is the right behaviour unless later we find the contrary is needed in the transformations)

And if you relaxed the `ancestor(Routine)` you can do next/previous_access from the two references in the initialisation expression, but not jump into the body of the subroutine. Which I also think is the right behaviour for now.

Reply:
It doesn't find in the initialisation expression.
I'll try to do add the same for the initialisation expression if I can work it out.

Reply:
@sergisiso Is this test sufficient for the non-routine access case or do you still want another standalone case?